test/TokenPaymaster.test.js

```diff
@@ -0,0 +1,228 @@
+      assert.equal((ethAfter - ethBefore) / 1e18, 2)
+    })
+
+    it('swap and transfer', async () => {
+      await token.mint(10e18.toString())
+      const target = '0x' + '1'.repeat(40)
+      const tokensBefore = await token.balanceOf(from)
+      const ethBefore = await web3.eth.getBalance(target)
+      await uniswap.tokenToEthTransferOutput(2e18.toString(), -1, -1, target)
+      const tokensAfter = await token.balanceOf(from)
+
+      const ethAfter = await web3.eth.getBalance(target)
+      assert.equal((tokensAfter - tokensBefore) / 1e18, -4)
+      assert.equal((ethAfter - ethBefore) / 1e18, 2)
+    })
+  })
+
+  context('#acceptRelayedCall', async () => {
+    it('fail if not enough balance', async () => {
+      const relayRequest = new RelayRequest({
+        ...sharedRelayRequestData
+      })
+      assert.deepEqual(await retcode(paymaster.acceptRelayedCall(relayRequest, '0x', 1e6)), {
+        code: 99,
+        msg: 'balance too low'
+      })
+    })
+
+    it('fund recipient', async () => {
+      await token.mint(5e18.toString())
+      await token.transfer(recipient.address, 5e18.toString())
+    })
+
+    it('fail if no approval', async () => {
+      const relayRequest = new RelayRequest({
+        ...sharedRelayRequestData
+      })
+      assert.deepEqual(await retcode(paymaster.acceptRelayedCall(relayRequest, '0x', 1e6)), {
+        code: 99,
+        msg: 'allowance too low'
+      })
+    })
+
+    it('recipient.approve', async () => {
+      await recipient.execute(token.address, token.contract.methods.approve(paymaster.address, -1).encodeABI())
+    })
+
+    it('succeed?', async () => {
+      const relayRequest = new RelayRequest({
+        ...sharedRelayRequestData
+      })
+      const ret = await retcode(paymaster.acceptRelayedCall(relayRequest, '0x', 1e6))
+      assert.equal(ret.code, 0, ret)
+    })
+  })
+
+  context('relayedCall', () => {
+    const paymasterDeposit = 1e18.toString()
+
+    before(async () => {
+      await hub.stake(relay, 7 * 24 * 3600, {
+        from: relayOwner,
+        value: 2e18
+      })
+      await hub.registerRelay(2e16.toString(), '10', 'url', { from: relay })
+      await hub.depositFor(paymaster.address, { value: paymasterDeposit })
+    })
+
+    it('pay with token to make a call', async () => {
+      const preTokens = await token.balanceOf(recipient.address)
+      const prePaymasterTokens = await token.balanceOf(paymaster.address)
+      // for simpler calculations: we don't take any fee, and gas price is '1', so actual charge
+      // should be exactly gas usage. token is 2:1 to eth, so we expect to pay exactly twice the "charge"
+      const relayRequest = new RelayRequest({
+        ...sharedRelayRequestData,
+        senderAddress: from,
+        senderNonce: (await hub.getNonce(from)).toString(),
+        gasPrice: '1',
+        pctRelayFee: '0',
+        baseRelayFee: '0'
+      })
+
+      const chainId = await web3.eth.net.getId()
```

Comment:
Conveniently and not confusingly at all, 'getID' will give you network ID. This is not chain ID. But both are '1' for ETH manner. 
There is a method to get chainID from web3, (```web3.eth.getChainId()```) but I use Environment class for that.

Reply:
There is no `web3.eth.getChainId()` accessible for me in tests or in truffle console...
I changed the code to use your Environments, even though its not relevant for ganache.

FWIW, 
`curl -X POST --data '{"jsonrpc":"2.0","method":"eth_chainId", "id":1}' localhost:8545`
results:
`{"id":1,"jsonrpc":"2.0","result":"0x539"}`

But otherwise, its irrelevant.